extensions/notification/NotificationQueuePopulator.js

```diff
@@ -119,22 +119,76 @@ class NotificationQueuePopulator extends QueuePopulatorExtension {
         return this.bnConfigManager.removeConfig(bucketName);
     }
 
+    /**
+     * Returns the correct versionId
+     * to display according to the
+     * versioning state of the object
+     * @param {Object} value log entry object
+     * @param {Object} overheadFields - extra fields missing from the log entry
+     * @return {String} versionId
+     */
+    _getVersionId(value, overheadFields) {
+        const versionId = value.versionId || (overheadFields && overheadFields.versionId);
+        const isNullVersion = value.isNull;
+        const isVersioned = !!versionId;
+        // Versioning suspended objects have
+        // a versionId, however it is internal
+        // and should not be used to get the object
+        if (isNullVersion || !isVersioned) {
+            return null;
+        }
+        return versionId;
+    }
+
+    /**
+     * Decides if we should process the entry.
+     * Since we get both master and version events,
+     * we need to avoid pushing two notifications for
+     * the same event.
+     * - For non-versioned buckets, we process the master
+     * objects' events.
+     * - For versioned buckets, we process version events
+     * and ignore all master events.
+     * - For versioning suspended buckets, we need to process
+     * both master and version events, as the master is considered
+     * a separate version.
+     * @param {String} key object key
+     * @param {Object} value object metadata
+     * @return {boolean} - true if entry is valid
+     */
+    _shouldProcessEntry(key, value) {
+        const isMaster = isMasterKey(key);
+        const hasVersionId = !!value.versionId;
+        const keyHasNullVersion = key.endsWith('\0');
+        const isVersionedMaster = isMaster && hasVersionId;
+        const isNullVersion = (isMaster || keyHasNullVersion) && value.isNull;
+        const isNullWithVersionId = hasVersionId && value.isNull && !keyHasNullVersion;
```

Comment:
minor points:

- `hasVersionId` is redundant because versions with `isNull` attribute always have a version ID.

- on the `keyHasNullVersion` variable name: I've used "null key" to refer to those keys ending with `\0` in Metadata, not claiming that this is the best name, but it could be better to stick with it for consistency (so naming the variable `isNullKey` instead of `keyHasNullVersion` which I also find confusing as "null version" has another meaning)